trinity/config.py

```diff
@@ -440,14 +452,60 @@ def nodekey(self, value: Union[bytes, PrivateKey]) -> None:
                 f"`PrivateKey` instance: got {type(self._nodekey)}"
             )
 
+    @contextmanager
+    def process_id_file(self, process_name: str):  # type: ignore
+        with PidFile(process_name, self.data_dir):
+            yield
+
     @classmethod
-    def from_parser_args(cls, parser_args: argparse.Namespace) -> 'TrinityConfig':
+    def from_parser_args(cls,
+                         parser_args: argparse.Namespace,
+                         sub_config_types: Iterable[Type['BaseSubConfig']]) -> 'BaseApplicationConfig':
         """
         Helper function for initializing from the namespace object produced by
         an ``argparse.ArgumentParser``
         """
         constructor_kwargs = construct_trinity_config_params(parser_args)
-        return cls(**constructor_kwargs)
+        trinity_config = cls(**constructor_kwargs)
+
+        trinity_config.initialize_sub_configs(parser_args, sub_config_types)
+
+        return trinity_config
+
+    def initialize_sub_configs(self,
+                               parser_args: argparse.Namespace,
+                               sub_config_types: Iterable[Type['BaseSubConfig']]) -> None:
+
+        for sub_config_type in sub_config_types:
+            self.add_sub_config(sub_config_type.from_parser_args(parser_args, self))
+
+    def add_sub_config(self, sub_config: 'BaseSubConfig') -> None:
+        self._sub_configs[type(sub_config)] = sub_config
+
+    def has_sub_config(self, sub_config_type: Type['BaseSubConfig']) -> bool:
+        return sub_config_type in self._sub_configs.keys()
+
+    def get_sub_config(self, sub_config_type: Type[TSubConfig]) -> TSubConfig:
+        # We want this API to return the specific type of the sub config that is requested.
+        # Our backing field only knows that it is holding `BaseSubConfig`'s but not concrete types
+        return cast(TSubConfig, self._sub_configs[sub_config_type])
+
+
+class BaseSubConfig:
+
+    base_config: BaseApplicationConfig = None
```

Comment:
@pipermerriam this feels a bit weird but maybe I was misunderstanding your proposal. Now I created these here because if we don't define the `__init__` how do we standardize where to access the `base_config` from sub configs?